server/src/main/java/org/jboss/as/server/deployment/module/ModuleDependency.java

```diff
@@ -71,7 +77,22 @@ public String toString() {
      * @param userSpecified {@code true} if this dependency was specified by the user, {@code false} if it was automatically added
      */
     public ModuleDependency(final ModuleLoader moduleLoader, final String identifier, final boolean optional, final boolean export, final boolean importServices, final boolean userSpecified) {
-        this(moduleLoader, ModuleIdentifier.create(identifier), optional, export, importServices, userSpecified);
+        this(moduleLoader, ModuleIdentifier.create(identifier), optional, export, importServices, userSpecified, Optional.empty());
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param moduleLoader the module loader of the dependency (if {@code null}, then use the default server module loader)
+     * @param identifier the module identifier
+     * @param optional {@code true} if this is an optional dependency
+     * @param export {@code true} if resources should be exported by default
+     * @param importServices
+     * @param userSpecified {@code true} if this dependency was specified by the user, {@code false} if it was automatically added
+     * @param reason reason for adding implicit module dependency
+     */
+    public ModuleDependency(final ModuleLoader moduleLoader, final String identifier, final boolean optional, final boolean export, final boolean importServices, final boolean userSpecified, Optional<String> reason) {
+        this(moduleLoader, ModuleIdentifier.create(identifier), optional, export, importServices, userSpecified, reason);
```

Comment:
Hi @khermano , don't use an Optional for a constructor.

The point is since this class already exposes two different constructors, the caller choosing one or another will be able to add the reason optionally.

We don't want the callers to be creating Optionals with a String. The caller will choose one of the overloaded constructors, the one without "String reason" to choose not reason at all, and the one with "String reason", to specify a reason.

Reply:
Oh, I am sorry I have misunderstood the previous comment.
So, if I understand it correctly now, the "reason" attribute should be a String, and just and only the getReason() method will be Optional<String>. Am I right?

Reply:
> Oh, I am sorry I have misunderstood the previous comment.

No problem at all!

> Am I right?

Yes, right, internally the class will use a String for the reason. This String can be null upon the kind of constructor involved, there won't be any setter, it can be final, and it will be exposed as an Optional<String> from the getter: if it is null, we should return an Empty Optional.

Reply:
Oh, good. Thank you.